Add Tile Classifier to enhance efficiency of tiny object detection

eugene123tw opened the pull request:
<!-- Contributing guide: https://github.com/openvinotoolkit/training_extensions/blob/develop/CONTRIBUTING.md -->

### Summary

<!--
Resolves #111 and #222.
Depends on #1000 (for series of dependent commits).

This PR introduces this capability to make the project better in this and that.

- Added this feature
- Removed that feature
- Fixed the problem #1234
-->

### How to test

<!-- Describe the testing procedure for reviewers, if changes are
not fully covered by unit tests or manual testing can be complicated. -->

### Checklist

<!-- Put an 'x' in all the boxes that apply -->

- [x] I submit my changes into the `develop` branch
- [x] I have added description of my changes into [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md)
- [ ] I have updated the [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) accordingly
- [x] I have added tests to cover my changes
- [x] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)

### License

- [x] I submit _my code changes_ under the same [MIT License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2023 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

eugene123tw committed:
add tile classifier

eugene123tw committed:
linter

eugene123tw committed:
add tile classifier - draft PR

eugene123tw committed:
code check changes

eugene123tw commented on otx/mpa/deploy/utils/onnx.py:
@cih9088 I had to add the following changes from mmdeploy, otherwise, there are dangling nodes in ONNX and OpenVINO would fail to partition my tile classifier model. I didn't dig too deep but it works, you might wanna have a look.

cih9088 replied:
LGTM

eugene123tw commented on otx/core/data/adapter/detection_dataset_adapter.py:
@sungmanc I only wanna polygon but I found this code includes boxes for instance segmentation and I end up having twice many of the annotations I wanted.

eugene123tw commented on otx/cli/utils/io.py:
@goodsong81 Do we need to change Geti backend since it has to read additional models? and who's responsible for this change?

goodsong81 replied:
As I said earlier, Geti does not use OTX CLI.
OTX should be responsible for
* Keep additional IR models using model.set_data()
* Extract, if any, additional IR models using model.get_data()
* Deploy, if any, additional IR models along with base one.
* Infer properly both in OpenVINO task and exportable code
* Keep backward compatibility for models deployed from early versions of OTX/Geti.

eugene123tw replied:
@goodsong81 My question pertains to the input of new models. Specifically, in the OTX CLI, we use the read_model function to read models from a folder. To add two new OpenVINO models, I need to include them with the other files that are required for the `model_adapter_keys variable`. These files include "tile_classifier.xml", "tile_classifier.bin". However, I am unsure where and how Geti reads these additional files before creating `task_environment` and initializing `OpenVINODetectionTask`. And there's nowhere I could test the Geti workflow (read models -> create task environment -> initialise OpenVINO task -> run prediction) behaviour from OTX, should we contact the Geti backend team for further information?

goodsong81 replied:
AFAIK, Geti will store models w/ their metadata given by set_data().
So, we may not need to worry.
Rather we are dealing with additional data in CLI use cases as we don't have such functionality.
You may test it after OTX is merged to Geti.

eugene123tw commented on otx/api/utils/tiler.py:
```diff
@@ -118,16 +122,25 @@ def predict_tile(
-        return features, output
+        run_prediction = True
+        if not return_features:
+            # TODO[EUGENE]: should we set the threshold (0.45) as a parameter?
```
@goodsong81 do you think we should set this confidence threshold as a configurable parameter? The higher the faster but risker.

goodsong81 replied:
Yeah, I think it's good to be included in config params with default values.

goodsong81 replied:
Hmm, or we just set the default good value for a while :)

eugene123tw replied:
then should we leave it at 0.45?

eugene123tw commented on otx/algorithms/common/tasks/training_base.py:
```diff
@@ -347,7 +347,7 @@ def _initialize_post_hook(self, options=None):
 
     @abc.abstractmethod
-    def _init_recipe(self):
+    def _init_recipe(self, dataset: Optional[DatasetEntity] = None):
```
@harimkang I added this parameter to a function to consolidate tiling-related parameters in one place instead of scattering them throughout the codebase. 

However, this change also affects other classes that inherit from this function, as the new parameter is now part of the function's signature and must be supplied when calling it. This requires updating the implementations of those classes to include the new parameter and its functionality.

harimkang replied:
So this is needed for tiling, so we need to add it to the BaseTask as well, right? I don't think there should be any major issues. If we add None as the default for the other tasks as well, shouldn't there be an issue?
@jaegukhyun What do you think?

jaegukhyun replied:
I think there is no harm for overall structure, but I think this could increase complexity of BaseTask, which is complex enough. Let's see... I'll consider how we can refactor for our new task. At the current status, there is no big objection

sungmanc replied:
Inserting `dataset` to `_init_recipe` function is not a good way for me since `recipe` is not related with dataset

How about using `_initialize` function?

eugene123tw committed:
quality fixes

sungmanc commented on otx/core/data/adapter/detection_dataset_adapter.py:
```diff
@@ -40,8 +41,10 @@ def get_otx_dataset(self) -> DatasetEntity:
                             and ann.type == AnnotationType.polygon
                         ):
-                            if self._is_normal_polygon(ann):
+                            if isinstance(ann, Polygon):
```
I'm not sure why `ann.type == AnnotationType.polygon` couldn't filter out `Bbox`.

If you want to use `isinstance(ann, Polygon)`, then edit above `if` sentence like below.
```python3
if (
    self.task_type in (TaskType.INSTANCE_SEGMENTATION, TaskType.ROTATED_DETECTION) and isinstance(ann, Polygon)
):
```

eugene123tw replied:
yeah, you're right. I reverted the change.

yunchu commented:
@eugene123tw could you set the milestone for this PR?

goodsong81 commented on otx/algorithms/detection/tasks/inference.py:
```diff
@@ -96,6 +97,47 @@ def __init__(self, task_environment: TaskEnvironment, **kwargs):
+                                )
+                            hparams.tiling_parameters.enable_tile_classifier = loaded_tiling_parameters[
+                                "enable_tile_classifier"
+                            ]["value"]
```
Need refactoring. Too much tiling-specific logics in _load_model_ckpt()

eugene123tw replied:
thanks for the comment, I've updated it.

goodsong81 reviewed (commented):
You are already doing OK w/ model.set/get_data()
Model will be saved & loaded this data from model repo.

eugene123tw committed:
quality checks

eugene123tw committed:
pylint fixes

eugene123tw committed:
Merge branch 'develop' into eugene/CVS-102976-tile-classifier

eugene123tw committed:
add description to changelog

eugene123tw committed:
update changelog

eugene123tw committed:
add config dict flatten func

eugene123tw committed:
fix tests

eugene123tw committed:
add tiling tests

eugene123tw committed:
add tests

eugene123tw committed:
run code quality checks

eugene123tw committed:
remove test

eugene123tw committed:
Merge branch 'develop' into eugene/CVS-102976-tile-classifier

eugene123tw committed:
fix conflicts

eugene123tw committed:
quality improvement

eugene123tw committed:
update

eugene123tw committed:
fix pylint

eugene123tw committed:
add pylint disable

eugene123tw committed:
make mypy happy

eugene123tw committed:
try to make mypy happy

jaegukhyun reviewed (commented):
I don't have any big objection for this PR, but I want to check integration test and e2e test for all tasks. @eugene123tw Could you attach your local e2e test for all tasks?

eugene123tw committed:
update for mypy

eugene123tw committed:
make mypy happy

eugene123tw committed:
add fixes

eugene123tw committed:
update

eugene123tw committed:
Merge branch 'eugene/CVS-102976-tile-classifier' of https://github.com/eugene123tw/training_extensions into eugene/CVS-102976-tile-classifier

eugene123tw committed:
fix onnx partition

eugene123tw committed:
quality checks

eugene123tw committed:
fix deployment

eugene123tw committed:
fix tests

eugene123tw committed:
fix tests

eugene123tw committed:
fix errors

eugene123tw committed:
fix too many branches issue in io.py

eugene123tw committed:
fix io

eugene123tw committed:
fix explainer

eugene123tw committed:
add explain export

eugene123tw commented:
@jaegukhyun pre-merge tests and CLI tests finished, and codecov also attached. Let me know if there's any info you need. Thanks!

eugene123tw committed:
add tests

jaegukhyun commented on otx/algorithms/detection/adapters/mmdet/utils/config_utils.py:
```diff
@@ -326,3 +327,58 @@ def cluster_anchors(recipe_config: Config, dataset: DatasetEntity):
+            config.model.type = "CustomMaskRCNNTileOptimised"
+
+            if config.model.backbone.type == "efficientnet_b2b":
+                lr = 0.002  # pylint: disable=invalid-name
```
```suggestion
                learning_rate = 0.002
```

jaegukhyun commented on otx/algorithms/detection/adapters/mmdet/utils/config_utils.py:
```diff
@@ -326,3 +327,58 @@ def cluster_anchors(recipe_config: Config, dataset: DatasetEntity):
+            if config.model.backbone.type == "efficientnet_b2b":
+                lr = 0.002  # pylint: disable=invalid-name
+                logger.info(
+                    f"Patch {config.model.backbone.type} LR from: {hparams.learning_parameters.learning_rate} to {lr}"
```
```suggestion
                    f"Patch {config.model.backbone.type} LR from: {hparams.learning_parameters.learning_rate} to {learning_rate}"
```

jaegukhyun commented on otx/algorithms/detection/adapters/mmdet/utils/config_utils.py:
```diff
@@ -326,3 +327,58 @@ def cluster_anchors(recipe_config: Config, dataset: DatasetEntity):
+                logger.info(
+                    f"Patch {config.model.backbone.type} LR from: {hparams.learning_parameters.learning_rate} to {lr}"
+                )
+                hparams.learning_parameters.learning_rate = lr
```
```suggestion
                hparams.learning_parameters.learning_rate = learning_rate
```

jaegukhyun commented on otx/algorithms/detection/adapters/mmdet/utils/config_utils.py:
```diff
@@ -326,3 +327,58 @@ def cluster_anchors(recipe_config: Config, dataset: DatasetEntity):
+                overlap_ratio=float(hparams.tiling_parameters.tile_overlap),
+                max_per_img=int(hparams.tiling_parameters.tile_max_number),
+            )
+            config.update(
```
What if hparams.tiling_parameters.enable_adaptive_param is False? Is there another logic to add default tiling_params to config or there is no need to add tiling_params to config in that case?

eugene123tw replied:
If enable adaptive param is set to false then the default tiling parameters is used:  https://github.com/openvinotoolkit/training_extensions/blob/fd1e6c3cd27c35add8cae76ab474cffd9d5e0c5d/otx/algorithms/detection/configs/detection/configuration.yaml#L411

jaegukhyun replied:
Yes but those default parameters are updated to config? In this code config is updated only when adaptive tiling params is true

eugene123tw replied:
Tiling parameters already exist in the config as `tile_pipeline.py` is being used instead of using `data_pipeline` https://github.com/openvinotoolkit/training_extensions/blob/5ae47128adc6d312d2b0ae1d29e8e1c0e1a1f5ae/otx/algorithms/common/tasks/training_base.py#L213

jaegukhyun commented on otx/algorithms/detection/adapters/mmdet/utils/config_utils.py:
```diff
@@ -326,3 +327,58 @@ def cluster_anchors(recipe_config: Config, dataset: DatasetEntity):
+                )
+            )
+
+        if dataset and dataset.purpose == DatasetPurpose.INFERENCE:
```
Don't we have setting samples_per_gpu to 1 for test_dataloader? If we don't have that logic just resolve this. I think I saw that logic in the stage

eugene123tw replied:
It is set to 2 (or 4, I can't remember) in our current code, although I don't know where it is set yet. 

Before I change this universally to 1, I'd like to confirm check the batch size should be set to 1? I think this is set to 1 before, but somehow set to > 1 and I thought this is intentional, no?

jaegukhyun replied:
Okay, then we should set validation batch size in the task level, I'll keep this in my mind for refactoring

jaegukhyun reviewed (commented):
I left some comments, could you check them?

eugene123tw committed:
Apply suggestions from code review

Co-authored-by: Jaeguk Hyun <jaeguk.hyun@intel.com>

eugene123tw committed:
revert datumaro changes

eugene123tw committed:
pyline changes

eugene123tw committed:
solve conflict

eugene123tw committed:
solve conflicts

change onnx.py to onnx_utils.py avoiding import error

change onnx.py to onnx_utils.py avoiding import error

try pylint again

try again

eugene123tw committed:
disable pylint

eugene123tw committed:
fix tests

sungmanc commented on otx/algorithms/detection/adapters/mmdet/models/detectors/custom_maskrcnn_tile_optimised.py:
```diff
@@ -0,0 +1,312 @@
+"""Custom MaskRCNN detector with tile classifier."""
```
Most of contents included in this PR used American style word. So, want to suggest to change the title.
optimised --> optimized

eugene123tw replied:
Done

sungmanc commented on otx/algorithms/detection/adapters/mmdet/models/detectors/custom_maskrcnn_tile_optimised.py:
```diff
@@ -0,0 +1,312 @@
+        super().__init__()
+        self.fp16_enabled = False
+        self.features = nn.Sequential(
+            nn.Conv2d(3, 64, kernel_size=11, stride=4, padding=2),
```
Could you modularize this part? maybe we can use below style function

```python
def make_conv_block(in_channels, out_channels, k_size, stride, padding) -> nn.Sequential:
```

eugene123tw replied:
I have changed this part. Although, not sure if it's the style you referred to.

sungmanc commented on otx/algorithms/detection/adapters/mmdet/models/detectors/custom_maskrcnn_tile_optimised.py:
```diff
@@ -0,0 +1,312 @@
+            torch.nn.Linear(256, 1),
+        )
+
+        self.loss_fun = torch.nn.BCEWithLogitsLoss()
```
```suggestion
        self.loss_func = torch.nn.BCEWithLogitsLoss()
```

sungmanc commented on otx/algorithms/detection/adapters/mmdet/models/detectors/custom_maskrcnn_tile_optimised.py:
```diff
@@ -0,0 +1,312 @@
+        Returns:
+            torch.Tensor: BCE loss
+        """
+        loss = self.loss_fun(pred, target)
```
```suggestion
        loss = self.loss_func(pred, target)
```

sungmanc commented on otx/algorithms/detection/adapters/mmdet/utils/config_utils.py:
```diff
@@ -326,3 +327,59 @@ def cluster_anchors(recipe_config: Config, dataset: DatasetEntity):
+            logger.info(f"Patch model from: {config.model.type} to CustomMaskRCNNTileOptimised")
+            config.model.type = "CustomMaskRCNNTileOptimised"
+
+            if config.model.backbone.type == "efficientnet_b2b":
```
This patching seems to be avoided. It is easy to miss that the learning rate is changed.
Could we make this part to be more explicit way?

sungmanc commented on otx/api/configuration/helper/utils.py:
```diff
@@ -187,3 +187,32 @@ def config_to_bytes(config: ConfigurableParameters) -> bytes:
+                flatten_config_values(value)
+
+
+def group_to_dict(config: dict):
```
If this function only targets the detection task, how about changing the name?

eugene123tw replied:
I renamed it.

sungmanc reviewed (commented):
To be honest, I couldn't fully understand the logic. BTW, I left some comments

eugene123tw committed:
rename file

eugene123tw committed:
address comments

eugene123tw committed:
sort imports

jaegukhyun reviewed (changes requested):
https://github.com/openvinotoolkit/training_extensions/pull/1955 Overall detection task is refactored. You need to move your PR's content to new task structure. I know this is quite annoying, but I hope you understand this situation.

eugene123tw committed:
Merge branch 'develop' into eugene/CVS-102976-tile-classifier

eugene123tw committed:
fix conflicts

eugene123tw committed:
quality checks

eugene123tw committed:
fix test

jaegukhyun reviewed (changes requested):
Detection task test took so many long time. Could you check it in local?

eugene123tw committed:
disable otx hpo test

eugene123tw committed:
refactor

jaegukhyun commented on otx/algorithms/detection/adapters/mmdet/task.py:
```diff
@@ -199,6 +200,7 @@ def _init_task(self, dataset: Optional[DatasetEntity] = None, export: bool = Fal
         # Patch tiling parameters
         patch_tiling(self._recipe_cfg, self._hyperparams, dataset)
 
+        set_hyperparams(self._recipe_cfg, self._hyperparams)
```
Do we need this? Hyperparameters are updated in line#169.

eugene123tw replied:
Thanks for letting me know! I added some changes.

jaegukhyun reviewed (commented):
Why don't you remove unnecessary changes after detection task refactoring?

eugene123tw committed:
remove set_params

eugene123tw committed:
import fix

eugene123tw committed:
Merge branch 'develop' into eugene/CVS-102976-tile-classifier

eugene123tw committed:
solve mypy issue

eugene123tw committed:
fix test

eugene123tw committed:
reinstate hpo test

eugene123tw committed:
fix tests

jaegukhyun commented on otx/algorithms/detection/task.py:
```diff
@@ -98,7 +98,44 @@ def __init__(self, task_environment: TaskEnvironment, output_path: Optional[str]
+            model (Optional[ModelEntity]): The model entity.
+
+        Returns:
+            _type_: _description_
```
Could you check this description?

eugene123tw replied:
Done.

jaegukhyun reviewed (commented):
I left one comment!

eugene123tw committed:
add method description

eugene123tw committed:
Merge branch 'develop' into eugene/CVS-102976-tile-classifier

eugene123tw committed:
add resume to integration tiling test

eugene123tw committed:
skip hpo test

eugene123tw committed:
fix feature vector and saliency map output in async

sungmanc reviewed (changes requested):
Need to resolve conflicts.
BTW, AFAIK, you have been experimenting with a lower learning rate to avoid the patching.
If the experimental score is quite good, will you introduce a new learning rate in this PR?

eugene123tw committed:
Merge branch 'develop' into eugene/CVS-102976-tile-classifier

eugene123tw committed:
Merge branch 'develop' into eugene/CVS-102976-tile-classifier

eugene123tw committed:
solve conflicts

eugene123tw committed:
make mypy happy

eugene123tw committed:
rename tiling test

eugene123tw committed:
add integration test

eugene123tw committed:
improve docstring

supersoob commented on otx/algorithms/detection/task.py:
```diff
@@ -290,6 +320,20 @@ def export(
+
+        if self._hyperparams.tiling_parameters.enable_tile_classifier:
+            tile_classifier = None
+            for partition in outputs.get("partitioned"):
```
Minor, but suggest -> outputs.get("partitioned", {})

eugene123tw replied:
Thanks! That's a great catch!

supersoob commented on otx/api/utils/tiler.py:
```diff
@@ -68,87 +100,98 @@ def predict(self, image: np.ndarray):
         """
+        tile_coords = self.tile(image)
+        if isinstance(self.classifier, Model):
+            tile_coords = self.filter_tiles_by_objectness(image, tile_coords)
```
I think these lines are duplicated with `predict_sync`. Filtering tiles from tiling classifier uses `infer_sync`, which doesn't seems appropriate inside `predict_async`. I don't know well but if possible, it'd be better to move them into `predict` and then branch off into each `predict_sync` and `predict_async` with input of tile_coords. If not possible and hard, it's fine to let as is.

eugene123tw replied:
Thank you for your feedback. I agree that it'd be ideal to have the code structured in the way you suggested, but as you mentioned, it may not be practical to implement at this time and I need some time to think about the overall design. I appreciate your suggestion and I will add a TODO in the code to revisit this in the future when feasible.

eugene123tw replied:
Thanks! That's great advice. I've refactored accordingly.

supersoob reviewed (commented):
Nice work:) Thanks for your efforts. I have minor comments.

jaegukhyun commented on tests/integration/cli/detection/test_tiling_instseg.py:
```diff
@@ -34,11 +37,28 @@
         "4",
         "--tiling_parameters.enable_tiling",
         "1",
+        "--tiling_parameters.enable_tile_classifier",
```
Shouldn't we check the non-tile-classifier cli test?

eugene123tw replied:
I could run additional testing, but doing so would double the testing time. My reasoning is that since enabling the tile classifier already covers the tiling test, running extra tests is unnecessary.

eugene123tw committed:
address comment

eugene123tw committed:
Merge branch 'develop' into eugene/CVS-102976-tile-classifier

eugene123tw commented:
> Need to resolve conflicts. BTW, AFAIK, you have been experimenting with a lower learning rate to avoid the patching. If the experimental score is quite good, will you introduce a new learning rate in this PR?

I have conducted a study comparing the learning rate on EfficientNet for the Vitens Coliform dataset. The results indicate a significant 5% accuracy gap between the different learning rates. It's worth noting that the training on the lower learning rate was terminated earlier due to early stop. While this outcome is not surprising, I believe that the 5% gap is substantial enough to keep the LR patching approach.

eugene123tw committed:
address soob's comment

goodsong81 reviewed (approved):
LTGM! Thank you for the great feature & enhancement.

sungmanc reviewed (approved):
Thanks for the work
